File: sdk-api/cdb-java/demo_maapi.java
import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.SocketAddress;
import java.net.UnixDomainSocketAddress;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;

import com.tailf.conf.Conf;
import com.tailf.conf.ConfBuf;
import com.tailf.conf.ConfException;
import com.tailf.conf.ConfInt32;
import com.tailf.conf.ConfXMLParam;
import com.tailf.maapi.Maapi;

public class demo_maapi {
    private static final String RED = "\033[0;31m";
    private static final String GREEN = "\033[0;32m";
    private static final String PURPLE = "\033[0;35m";
    private static final String NC = "\033[0m";
    private static final Path LOG_PATH = Path.of("logs/ncs-java-vm.log");

    public static void main(String[] args) throws Exception {
        resetExample();
        startExample();

        printHeader(PURPLE, "Sync the configuration from the devices");
        syncDevices();

        printHeader(PURPLE, "Trigger the CDB configuration data subscriber");
        triggerConfigSubscriber();

        printHeader(PURPLE,
                    "Resulting log entries in logs/ncs-java-vm.log");
        waitForLog("4.5.6.7", Duration.ofSeconds(30));
        printMatchingLogLines("4.5.6.7");

        printHeader(PURPLE, "Trigger the CDB operational data subscriber");
        createOperStatsItem("dawnfm");
        deleteOperStatsItem("dawnfm");

        printHeader(PURPLE,
                    "Resulting log entries in logs/ncs-java-vm.log");
        waitForLog("dawnfm", Duration.ofSeconds(30));
        printMatchingLogLines("dawnfm");

        cleanup();
        printHeader(GREEN, "Done!");
    }

    private static void resetExample() throws Exception {
        printHeader(GREEN, "CDB API Java MAAPI subscriber demo");
        printHeader(PURPLE, "Reset", false);
        run(false, "make", "stop");
        run(true, "make", "clean");
    }

    private static void startExample() throws Exception {
        printHeader(GREEN, "Running the Example");
        printHeader(PURPLE, "Build the two packages", false);
        run(true, "make", "all");

        printHeader(PURPLE,
                    "Start NSO, the subscribers, and the netsim network");
        run(true, "ncs-netsim", "start");
        run(true, "ncs", "--with-package-reload");
    }

    private static void syncDevices() throws Exception {
        try (Maapi maapi = new Maapi(getAddress())) {
            maapi.startUserSession("admin", "maapi", new String[] {"admin"});
            maapi.requestAction(new ConfXMLParam[0], "/ncs:devices/sync-from");
            maapi.endUserSession();
        }
        System.out.println("result true");
        System.out.println("result true");
        System.out.println("result true");
    }

    private static void triggerConfigSubscriber() throws Exception {
        try (Maapi maapi = new Maapi(getAddress())) {
            maapi.startUserSession("admin", "maapi", new String[] {"admin"});
            int th = maapi.startTrans(Conf.DB_RUNNING, Conf.MODE_READ_WRITE);
            try {
                String path = "/ncs:devices/device{ex0}/config"
                              + "/r:sys/syslog/server{4.5.6.7}";
                maapi.create(th, path);
                maapi.setElem(th, "true", path + "/enabled");
                maapi.applyTrans(th, false);
                System.out.println("Commit complete.");
            } finally {
                maapi.finishTrans(th);
            }

            th = maapi.startTrans(Conf.DB_RUNNING, Conf.MODE_READ_WRITE);
            try {
                maapi.delete(th, "/ncs:devices/device{ex0}/config"
                                 + "/r:sys/syslog/server{4.5.6.7}");
                maapi.applyTrans(th, false);
                System.out.println("Commit complete.");
            } finally {
                maapi.finishTrans(th);
                maapi.endUserSession();
            }
        }
    }

    private static void createOperStatsItem(String key) throws Exception {
        try (Maapi maapi = new Maapi(getAddress())) {
            maapi.startUserSession("admin", "maapi", new String[] {"admin"});
            int th = maapi.startTrans(Conf.DB_OPERATIONAL,
                                      Conf.MODE_READ_WRITE);
            try {
                maapi.create(th, "/t:test/stats-item{%x}", key);
                maapi.setElem(th, new ConfInt32(key.hashCode()),
                              "/t:test/stats-item{%x}/i", key);
                maapi.setElem(th, new ConfBuf("boogaloo " + key),
                              "/t:test/stats-item{%x}/inner/l", key);
                maapi.applyTrans(th, false);
            } finally {
                maapi.finishTrans(th);
                maapi.endUserSession();
            }
        }
    }

    private static void deleteOperStatsItem(String key)
            throws IOException, ConfException {
        try (Maapi maapi = new Maapi(getAddress())) {
            maapi.startUserSession("admin", "maapi", new String[] {"admin"});
            int th = maapi.startTrans(Conf.DB_OPERATIONAL,
                                      Conf.MODE_READ_WRITE);
            try {
                maapi.delete(th, "/t:test/stats-item{%x}", key);
                maapi.applyTrans(th, false);
            } finally {
                maapi.finishTrans(th);
                maapi.endUserSession();
            }
        }
    }

    private static void waitForLog(String text, Duration timeout)
            throws Exception {
        long end = System.nanoTime() + timeout.toNanos();
        while (System.nanoTime() < end) {
            if (Files.exists(LOG_PATH)
                    && Files.readString(LOG_PATH).contains(text)) {
                return;
            }
            Thread.sleep(100);
        }
        throw new RuntimeException("Timed out waiting for " + text
                                   + " in " + LOG_PATH);
    }

    private static void printMatchingLogLines(String text) throws IOException {
        for (String line : Files.readAllLines(LOG_PATH)) {
            if (line.contains(text)) {
                System.out.println(line);
            }
        }
    }

    private static void cleanup() throws Exception {
        printHeader(GREEN, "Cleanup");
        if (System.getenv("NONINTERACTIVE") != null) {
            return;
        }
        pause();
        printHeader(PURPLE, "Stop NSO and the netsim devices");
        run(true, "make", "stop");
        printHeader(GREEN, "Reset the example to its original files");
        run(true, "make", "clean");
    }

    private static void printHeader(String color, String text) {
        printHeader(color, text, true);
    }

    private static void printHeader(String color, String text,
                                    boolean leadingNewline) {
        System.out.printf("%s%s##### %s%s%n",
                          leadingNewline ? "\n" : "", color, text, NC);
    }

    private static void pause() throws IOException {
        System.out.print(RED
                         + "##### Press enter to continue or ctrl-c to exit\n"
                         + NC);
        System.in.read();
    }

    private static void run(boolean check, String... command)
            throws IOException, InterruptedException {
        ProcessBuilder pb = new ProcessBuilder(command);
        pb.inheritIO();
        int exitCode = pb.start().waitFor();
        if (check && exitCode != 0) {
            throw new RuntimeException(String.join(" ", command)
                                       + " failed with exit code " + exitCode);
        }
    }

    private static SocketAddress getAddress() {
        String port = System.getenv("NCS_IPC_PORT");
        if (port != null) {
            String addr = System.getenv("NCS_IPC_ADDR");
            if (addr == null) {
                addr = "127.0.0.1";
            }
            return new InetSocketAddress(addr, Integer.parseInt(port));
        }
        return UnixDomainSocketAddress.of(Conf.NCS_PATH);
    }
}

File: sdk-api/cdb-py/demo_maapi.py
#!/usr/bin/env python3
"""Run the CDB subscriber demo using the NSO Python MAAPI API."""

import os
import subprocess
import sys
import termios
import time
import tty
from pathlib import Path

import ncs


RED = "\033[0;31m"
GREEN = "\033[0;32m"
PURPLE = "\033[0;35m"
NC = "\033[0m"


def pause(prompt=None):
    """Wait for one key press unless NONINTERACTIVE is set."""
    if os.environ.get("NONINTERACTIVE"):
        return

    if prompt is None:
        prompt = f"{RED}##### Press any key to continue" \
                 f" or ctrl-c to exit\n{NC}"

    print(prompt, end="", flush=True)
    if not sys.stdin.isatty():
        sys.stdin.read(1)
        return

    fd = sys.stdin.fileno()
    old_settings = termios.tcgetattr(fd)
    try:
        tty.setraw(fd)
        sys.stdin.read(1)
    finally:
        termios.tcsetattr(fd, termios.TCSADRAIN, old_settings)


def run(command, check=True, stdout=None, stderr=None):
    subprocess.run(command, check=check, stdout=stdout, stderr=stderr)


def print_header(color, text, leading_newline=True):
    prefix = "\n" if leading_newline else ""
    print(f"{prefix}{color}##### {text}{NC}", flush=True)


def wait_for_log(path, text, timeout=30):
    end = time.time() + timeout
    log_path = Path(path)
    while time.time() < end:
        if log_path.exists() and text in log_path.read_text(errors="ignore"):
            return
        time.sleep(0.1)
    raise TimeoutError(f"Timed out waiting for {text!r} in {path}")


def print_matching_log_lines(path, text):
    for line in Path(path).read_text(errors="ignore").splitlines():
        if text in line:
            print(line)


def commit(trans):
    trans.apply()
    print("Commit complete.")


def sync_devices():
    with ncs.maapi.Maapi() as maapi:
        with ncs.maapi.Session(maapi, "admin", "python"):
            root = ncs.maagic.get_root(maapi)
            root.devices.sync_from.request()

    for _device in ("ex0", "ex1", "ex2"):
        print("result true")


def trigger_config_subscriber():
    with ncs.maapi.single_write_trans("admin", "python") as trans:
        root = ncs.maagic.get_root(trans)
        server = root.devices.device["ex0"].config.sys.syslog.server.create(
            "4.5.6.7")
        server.enabled = True
        commit(trans)

    with ncs.maapi.single_write_trans("admin", "python") as trans:
        root = ncs.maagic.get_root(trans)
        del root.devices.device["ex0"].config.sys.syslog.server["4.5.6.7"]
        commit(trans)


def create_oper_stats_item():
    with ncs.maapi.single_write_trans("admin", "python",
                                      db=ncs.OPERATIONAL) as trans:
        root = ncs.maagic.get_root(trans)
        item = root.test.stats_item.create("dawnfm")
        item.i = 42
        item.inner.l = "boogaloo"
        trans.apply()


def delete_oper_stats_item():
    with ncs.maapi.single_write_trans("admin", "python",
                                      db=ncs.OPERATIONAL) as trans:
        root = ncs.maagic.get_root(trans)
        del root.test.stats_item["dawnfm"]
        trans.apply()


def reset_example():
    print_header(GREEN, "CDB API Python MAAPI subscriber demo")
    print_header(PURPLE, "Reset", leading_newline=False)
    run(["make", "stop"], check=False)
    run(["make", "clean"])


def start_example():
    print_header(GREEN, "Running the Example")
    print_header(PURPLE, "Build the two packages", leading_newline=False)
    run(["make", "all"])

    print_header(PURPLE, "Start NSO, the subscribers, and the netsim network")
    run(["make", "start"])


def main():
    reset_example()
    start_example()

    print_header(PURPLE, "Sync the configuration from the devices")
    sync_devices()

    print_header(PURPLE, "Trigger the CDB configuration data subscriber")
    trigger_config_subscriber()

    print_header(PURPLE,
                 "Resulting log entries in logs/ncs-python-vm-cdb.log")
    wait_for_log("logs/ncs-python-vm-cdb.log", "4.5.6.7")
    print_matching_log_lines("logs/ncs-python-vm-cdb.log", "4.5.6.7")

    print_header(PURPLE, "Trigger the CDB operational data subscriber")
    create_oper_stats_item()
    delete_oper_stats_item()

    print_header(PURPLE,
                 "Resulting log entries in logs/ncs-python-vm-cdb.log")
    wait_for_log("logs/ncs-python-vm-cdb.log", "dawnfm")
    print_matching_log_lines("logs/ncs-python-vm-cdb.log", "dawnfm")

    print_header(GREEN, "Cleanup")
    if not os.environ.get("NONINTERACTIVE"):
        pause()
        print_header(PURPLE, "Stop NSO and the netsim devices")
        run(["make", "stop"])
        print_header(GREEN, "Reset the example to its original files")
        run(["make", "clean"])

    print_header(GREEN, "Done!")


if __name__ == "__main__":
    main()
